### PR TITLE
Vehicle turret reloading

### LINKED ISSUE #17295: Setting turret mode (burst/semi) requires control access
Before, it could be done from the turret's position, which makes more sense for "primitive" turrets.

### LINKED ISSUE #17343: Turrets clone the shooter, preventing original shooter from being "responsible"
* Spawn a flaming electric car
* Get on a turret
* Spawn a NPC
* Commit murder

Observations:
* Committing the murder may be hard due to the cutting armor bug thing (not related to this bug)
* You don't use up moves - your clone does
* You don't gain skill. If you are at 99% ranged skill, you will get messages about skill level increasing every time you shoot. But it won't actually increase on you, just the clone
* You will most likely not be penalized for murder since `killer != &g->u`

Character clones shouldn't be used for anything that may have any lasting effects, only getting information.

### LINKED ISSUE #17618: Vehicle Guns do not accept ammo belts or magazines only  stacks of individual bullets or grenades.
As the title says my freshly unloaded Mark 19 Grenade launcher would not accept the ammo belt i just took out of it, just the grenades after disassembling the ammo belt.

### PR DESCRIPTION
**Pull request by mugling**
#### Summary

Turrets are now either reloaded by the player using ammo belts or automatically from their tanks. A number of outstanding bugs are fixed. Additionally there are some significant UI improvements
- Rebased from #17744
- Depends on #17732 (merged)
- Will include #17743
- Fixes #17343
- Fixes #17618
- Fixes #17295
#### Interface
- Reload or unload turrets by examining them
- Vehicle interaction display now shows `TURRET` and `CARGO` contents
- Targeting display now includes detail about amount and type of ammo
- Can switch turret mode (`semi`, `burst`, `auto` etc) via targeting UI
#### Implementation (JSON)
- Turrets are no longer `CARGO`, as a result seat + turret etc are now possible
- Vehicle definitions can specify instead percentage chance for `ammo`
- Existing save games are migrating by moving `CARGO` ammo to the turret `base` item
- Vehicle parts specify the `USE_TANKS` to support loading from vehicle `FUEL_TANK`
#### Implementation (C++)
- `turret_reload` and `turret_unload` are deprecated
- ~~A new `ranged` class is provided encapsulating an `item`~~
- The ~~derived~~ `turret_data` class overrides some methods, especially for `USE_TANKS` turrets
- Can call `ammo_set` on a turret
- Turret aiming is now separate from firing (the latter via `turret_data::fire`)
#### TODO

Marked CR as implementation needs review.
- [X] Implement unit tests à la #17456
- [X] Update mod content to new turret JSON format

### CODE REVIEW

**Comment by Coolthulhu** [comment]
> Will include #17456

You mean in this PR?

**Comment by mugling** [comment]
> Will include #17456
> 
> You mean in this PR?

Sorry no I mean #17746

**Comment by mugling** [comment]
f270079 includes a simplification of the design

File: src/game.cpp [hunk 1]
```diff
@@ -2830,13 +2830,41 @@ bool game::handle_action()
         case ACTION_FIRE:
             // Use vehicle turret or draw a pistol from a holster if unarmed
             if( !u.is_armed() ) {
+
                 int part = -1;
                 vehicle *veh = m.veh_at( u.pos(), part );
+
+                auto turret = veh->turret_query( u.pos() );
```
Comment (BevapDin) [inline src/game.cpp]: `veh` is not checked for validity here.

File: src/vehicle.h [hunk 2]
```diff
@@ -552,6 +616,12 @@ class vehicle : public JsonSerializer, public JsonDeserializer
     // Translate mount coords "p" into tile coords "q" using given pivot direction and anchor
     void coord_translate (int dir, const point &pivot, const point &p, point &q) const;
 
+    /** Get all vehicle parts (if any) at @ref pos optionally including @ref broken parts */
+    static std::vector<vehicle_part *> get_parts( const tripoint &pos, bool broken = false );
+
+    /** Get first part (if any) at @ref pos which matches the predicate @ref func */
+    static vehicle_part *get_part( const tripoint &pos, const std::function<bool(const vehicle_part *)>& func );
```
Comment (BevapDin) [inline src/vehicle.h]: Why does the callback work on a pointer? Does `get_part` ever pass a null pointer to the callback? Why does the actual callback (e.g. in `vehicle::turret_query`) not check for null? If the callback never gets a null pointer, it should take a reference, which implicitly promises that the argument will never be null.

**Comment by mugling** [comment]
#### Mods
* Tank mod and blazemod are updated to new JSON format
* We can use this PR combined with @Coolthulhu's #17746 to further cleanup blazemod


#### Unit tests
* Provided unit tests are kept intentionally simple as `catch.hpp` struggles with loops
* A fix is provided whereby unit tests crash when animation code runs

Ready for merge testing. Next PR in this series will include automatic turret targeting and possibly assigning NPC's as gunners.

File: tests/turrets.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                auto tank_idx = veh->install_part( 0, 0, tank->id, true );
+                REQUIRE( tank_idx >= 0 );
+                REQUIRE( veh->parts[ tank_idx ].ammo_set( default_ammo( ammo ) ) );
+
+            } else if( ammo ) {
+                veh->parts[ idx].ammo_set( default_ammo( ammo ) );
+            }
+
+            auto qry = veh->turret_query( veh->parts[ idx ] );
+			REQUIRE( qry );
```
Comment (Coolthulhu) [inline tests/turrets.cpp]: Tabs

**Comment by Coolthulhu** [comment]
M249 starts with a small drum magazine that isn't enough for 2 full bursts.
Even if it doesn't have much ammo, it should probably have big magazine.

**Comment by Coolthulhu** [comment]
Flamethrower refuses to burst fire, even if there are multiple targets.

**Comment by mugling** [comment]
M249 burst is too high, perhaps 25? Turrets installed by the player are not loaded upon installation. For predefined vehicles I'd prefer to implement a JSON option for specifying ammo type so that they could start with belts.

Either:
1. M249 burst becomes 25
2. M249 default magazine becomes 223 belt
3. Further PR adds option for specifying ammo type

I'd prefer (1) unless you think otherwise

**Comment by mugling** [comment]
> Flamethrower refuses to burst fire, even if there are multiple targets.

Are you aiming at a target or into empty space? IIRC all guns have that property whereby bursts are truncated if there are no targets. That behavior is of course wrong, but a pre-existing bug.

**Comment by Coolthulhu** [comment]
When switching modes of a humvee grenade launcher turret, it has 2 semi-auto (1) modes.

**Comment by mugling** [comment]
> That behavior is of course wrong, but a pre-existing bug.

Although I think I have a trivial fix that we could commit as part of this PR

**Comment by Coolthulhu** [comment]
> 1. M249 burst becomes 25
> 2. M249 default magazine becomes 223 belt

Both would be fine, but if 2. implies giving it 200 ammo by default, then 1 is better.

> Are you aiming at a target or into empty space?

A hostile target, with 3 other visible, hostile targets in range.

**Comment by mugling** [comment]
> That behavior is of course wrong, but a pre-existing bug.

Trivial fix is bfdc1cc and makes vehicle-mounted AOE weapons much more useful.

**Comment by Coolthulhu** [comment]
New reloading is a fair bit clunky. Would it be possible to make it sensible in this PR?
One of:
* With empty hands, when standing over a turret, make 'r' reload the turret rather than saying "you aren't wielding anything"
* Dropping a magazine on a turret would attempt reload instead of dropping. If reload fails, a message would be printed - "mag_name is not compatible with turret_name"

Also unloading the turret with 'U'. But that can wait.

**Comment by Coolthulhu** [comment]
> That behavior is of course wrong, but a pre-existing bug.

It's useful though. It makes burst conserve unused ammunition, which is pretty important.

**Comment by mugling** [comment]
> When switching modes of a humvee grenade launcher turret, it has 2 semi-auto (1) modes.

Well spotted, fixed by 3966c1d

**Comment by mugling** [comment]
> It's useful though. It makes burst conserve unused ammunition, which is pretty important.

The new fix has the correct behavior - if you're aiming into empty space fire the full burst. If you're aiming at a target then stop the burst if you run out of targets

**Comment by mugling** [comment]
> New reloading is a fair bit clunky. Would it be possible to make it sensible in this PR?

Fair enough, have pushed update

**Comment by mugling** [comment]
Fixed burst fire when using vehicle tanks

**Comment by mugling** [comment]
a8dd6ec ensures that stock vehicles come loaded with standard 40mm grenades

File: data/json/items/ammo_types.json [hunk 4]
```diff
@@ -159,7 +159,7 @@
         "type" : "ammunition_type",
         "id" : "40mm",
         "name" : "40mm grenade",
-        "default" : "40mm_concussive"
+        "default" : "40mm_grenade"
```
Comment (Coolthulhu) [inline data/json/items/ammo_types.json]: Doesn't exist.
You mean "frag"?

Comment (mugling) [inline data/json/items/ammo_types.json]: yes, fixed

**Comment by Coolthulhu** [comment]
Another important thing: reloading loaded turret.
Reloading a held weapon with a ready mag allows replacing and replenishing. Having to unload and reload is a lot of keypresses.

**Comment by mugling** [comment]
Done

**Comment by Coolthulhu** [comment]
* Spawn flaming electric car
* Spawn 10 STANAG drum magazines
* Reload the M249 with one
* debugmsg, one of the mags vanishes, none is placed in the inventory

**Comment by mugling** [comment]
Before or after latest commit?

**Comment by mugling** [comment]
And what debugmsg

**Comment by Coolthulhu** [comment]
After.
Debugmsg complains that it can't find a magazine to remove.
I'll try with different mag types to see if it fails to remove the old one or the new one.

**Comment by Coolthulhu** [comment]
It's quite weird, looks like some sort of function call ordering problem:
* Unload the turret
* Reload it with empty mag
* 2 debugmsgs - first one that it can't remove, second one that it tried to remove something not contained in the item (from vistitable.cpp)
* Reload the turret with ammo belt
* It asks me where to dispose the ammo belt (note: turret has a drum mag, I'm attaching the ammo belt)
* Debugmsg after that

**Comment by mugling** [comment]
Found it, fix is non-trivial so I'm going to revert fd4d6d1 for now

**Comment by mugling** [comment]
The problem is that `item::reload` uses `player::dispose_item` to discard magazines. The latter function was only designed to work with items in the players possession. A correct fix will require use of `item_location` but that's well beyond the scope of this PR.

**Comment by Coolthulhu** [comment]
That new reloading mechanic is considerably clunkier than old one. Having to manually unload and reload each turret is tedious even with just core turrets. Mods make heavy use of low capacity turrets.

Would it be possible to restore cargo turrets until that problem is solved? Changes would be confined just to a bunch of test turrets. This would prevent future conflicts and get the code in, but wouldn't kill the current turrets.

Because unfortunately, the current state of this PR is unacceptable in mainline due to the UI issues.

And a more minor issue:
Internal mag turrets, like in tank mod and many blazemod turrets, are not reloadable.

**Comment by mugling** [comment]
> Would it be possible to restore cargo turrets until that problem is solved?

No. It would require a near rewrite.

> Because unfortunately, the current state of this PR is unacceptable in mainline due to the UI issues.

The only issue being you can't part reload a turret? A lot of UI improvements were also made including the ability to display remaining turret ammo from the vehicle interaction screen along with other earlier changes like displaying the loaded ammo and mode in the targeting UI. Overall I think it's definitely a nett improvement.

Given this PR blocks the blazemod refactor I'd prefer to merge as-is and then provide a fix for `player::dispose_item`. If it's absolutely essential to you then I could kludge dispose_item via a check for the `VEHICLE` flag but this seems somewhat extreme considering the above.
 
> And a more minor issue:
> Internal mag turrets, like in tank mod and many blazemod turrets, are not reloadable.

Internal magazine turrets do work just not in those mods.I'd obviously prefer to fix blazemod as part of #17456 as it's otherwise excessive duplication of labor.

**Comment by Coolthulhu** [comment]
> The only issue being you can't part reload a turret?

* Have to manually unload a turret before reloading it
* Reloading inaccessible turrets takes more keypresses
* Severely limited turret capacity for "physical" turrets which already were inferior to "energy" ones means more reloading, which is now much clunkier than before
* Muscle memory for turret reloading no longer works
* Advanced inventory menu can no longer reload a turret

> Overall I think it's definitely a nett improvement.

For you, but I'm nearly certain that players who use turrets a lot will disagree.

> If it's absolutely essential to you then I could kludge dispose_item via a check for the VEHICLE flag but this seems somewhat extreme considering the above.

I'd prefer that option. A working hack is better than a possible future fix.
Alternatively you can make a PR on top of this one with relevant fixes and I'll avoid merging all PRs that conflict with this one (or the new one) until it's merged or left unupdated for 3-4 days or so.

> Internal magazine turrets do work just not in those mods.

Is this a regression introduced in this or earlier PR?

**Comment by mugling** [comment]
> Have to manually unload a turret before reloading it
> For you, but I'm nearly certain that players who use turrets a lot will disagree.

Does need fixing but I'm not sure it's critical compared to the fact that this PR allows unit testing of turrets and therefore brings us a lot closer to merging blazemod which would likely be very popular indeed.

> Reloading inaccessible turrets takes more keypresses
> Severely limited turret capacity for "physical" turrets which already were inferior to "energy" ones means more reloading, which is now much clunkier than before

Implicit consequences of  #17618 although longer ammo belts would be entirely reasonable and trivial to implement. How about doubling their maximum capacity. Ammo scarcity is a pre-existing problem.

> Muscle memory for turret reloading no longer works

This applies to all new features. No way around this.

> Advanced inventory menu can no longer reload a turret

Didn't even know it could. Will look into but hardly critical for this PR.

**Comment by Coolthulhu** [comment]
> Does need fixing but I'm not sure it's critical compared to the fact that this PR allows unit testing of turrets and therefore brings us a lot closer to merging blazemod which would likely be very popular indeed.

Changes to code that cause regressions to the current features can wait for PRs that build on them.

You can leave this PR open and build a new one on it and then I'll merge both.

> This applies to all new features. No way around this.

It would be possible to check when a compatible magazine is dropped on a turret and reload instead of having it fall through.

> Will look into but hardly critical for this PR.

One convenient method of reloading a turret is a must. I won't change my mind here.
You can drop code quality and introduce hacks to be removed later, make turret reloading not cost any moves and just use a simple item swap function rather than proper reload etc.
But reloading has to be convenient and must not involve manual unloads just for regular reloads or partial reloads ("top off" partial reload that use all ammo they can, limited reloads don't need to work).

Not being able to use all the convenient ammo drop options to reload turrets is a significant regression that, if not fixed, WILL result in players being frustrated at extra tedium that comes with, from player perspective, only very minor additions.

**Comment by mugling** [comment]
34626e6 allows for the same interface as reloading player weapons (partial reloading, limited reload, exchange of magazines). It extends `player::dispose_item` to expect an `item_location` and then provides the necessary kludge in `item:::reload`

**Comment by mugling** [comment]
d24431a allows 500 rounds for ammo belts

**Comment by Coolthulhu** [comment]
`item_location.cpp` needs `#include <algorithm>`.

**Comment by mugling** [comment]
> item_location.cpp needs #include <algorithm>.

Just noticed that. See also pushed fix for tank and blazemod so that turrets ignore `NO_RELOAD`

**Comment by Coolthulhu** [comment]
Useful things to do:
* Allow unloading nearby turrets with 'U'nload, if possible
* Capture drop attempt on turret, check if it is a magazine, try to reload if it is
* Something for advanced inventory menu, preferably like the above

**Comment by mugling** [comment]
> Allow unloading nearby turrets with 'U'nload, if possible

@codemime how hard would it be to extend the inventory UI to support turrets?

**Pull request merged**